### PR TITLE
Add convergence check

### PR DESCRIPTION
**Pull request by bantin**
Right now, all algorithms will share the same basic convergence check, which simply checks that the loss has stopped decreasing. In the future, we might want different convergence checks for different algorithms.

Addresses issue #36

### CODE REVIEW

File: src/model.jl [hunk 1]
```diff
@@ -97,12 +100,38 @@ function fit_cnmf(data; L=10, K=5, alg=:mult,
         # Record time and loss
         push!(time_hist, time_hist[end] + dur)
         push!(loss_hist, loss)
+
+        # Check convergence
+        if check_convergence && converged(loss_hist, patience, tol)
```
Comment (degleris1) [inline src/model.jl:105]: Any reason you chose to have this as a `break` statement instead of making it a condition for the `while` loop?

File: src/model.jl [hunk 2]
```diff
@@ -97,12 +100,38 @@ function fit_cnmf(data; L=10, K=5, alg=:mult,
...
 end
 
+"""
+Check for model convergence
+"""
+function converged(loss_hist, patience, tol)
+
+    # If we have not run for `patience` iterations,
+    # we have not converged.
+    if length(loss_hist) <= patience
```
Comment (degleris1) [inline src/model.jl:121]: Is this the best application of `patience`? My thought would be to use it for something like the number of iterations `d_loss` has to be small before exiting (`d_loss` would have to be small for 3 consecutive iterations for example).

**Review by degleris1 — COMMENTED** [review]
Looks good overall. See comments. Merge whenever you're ready.

File: src/model.jl (hunk 2 above)
Comment (bantin) [reply src/model.jl:121]: Yeah that's how I think it currently works. Looking at the code below:

```julia 
d_loss = diff(loss_hist[end-patience:end])

     # Objective converged
    if (all(abs.(d_loss) .< tol))
        return true
    else
        return false
    end
```
we take the first difference across `patience` number of prior loss values and check that they're all small. The check above is because when the array has fewer than `patience` number of entries  the below code would throw an indexing error.

File: src/model.jl (hunk 1 above)
Comment (bantin) [reply src/model.jl:105]: No good reason -- just seemed more awkward to have it in the while loop.  The logic of the while loop would have gotten a bit bulky (because `check_convergence` is optional) so this seemed more readable.
